File: algorithms.py
# algorithms.py

def calculate_scheduling(algorithm, processes, time_quantum=2):
    # processes: list of (pid, arrival, burst, priority)
    if algorithm == "FCFS":
        return fcfs(processes)
    elif algorithm == "SJF":
        return sjf(processes)
    elif algorithm == "Round Robin":
        return round_robin(processes, time_quantum)
    elif algorithm == "Priority Scheduling":
        return priority_scheduling(processes)
    return [], {}

def fcfs(processes):
    processes.sort(key=lambda x: x[1])  # Sort by Arrival Time
    current_time = 0
    segments = []
    pid_metrics = {}

    for pid, arrival, burst, priority in processes:
        if current_time < arrival:
            current_time = arrival
        
        start_time = current_time
        completion_time = start_time + burst
        waiting_time = start_time - arrival
        turnaround_time = completion_time - arrival
        
        segments.append((pid, start_time, completion_time))
        pid_metrics[pid] = {'wt': waiting_time, 'tat': turnaround_time}
        current_time = completion_time

    return segments, pid_metrics

def sjf(processes):
    # Shortest Job First (Non-preemptive)
    # Sort by Arrival, then Burst Time
    processes.sort(key=lambda x: (x[1], x[2]))
    remaining = processes.copy()
    current_time = 0
    segments = []
    pid_metrics = {}

    while remaining:
        available = [p for p in remaining if p[1] <= current_time]
        if not available:
            # Advance time if no process has arrived
            current_time = min(p[1] for p in remaining)
            continue
        
        # Pick process with shortest burst time
        available.sort(key=lambda x: x[2])
        process = available[0]
        remaining.remove(process)
        
        pid, arrival, burst, priority = process
        start_time = current_time
        completion_time = start_time + burst
        waiting_time = start_time - arrival
        turnaround_time = completion_time - arrival
        
        segments.append((pid, start_time, completion_time))
        pid_metrics[pid] = {'wt': waiting_time, 'tat': turnaround_time}
        current_time = completion_time

    return segments, pid_metrics

def round_robin(processes, quantum):
    processes.sort(key=lambda x: x[1])
    remaining = processes.copy()
    ready_queue = []
    current_time = 0
    segments = []
    pid_metrics = {}
    
    # Store remaining burst and original burst
    burst_remaining = {p[0]: p[2] for p in processes}
    arrival_times = {p[0]: p[1] for p in processes}
    original_bursts = {p[0]: p[2] for p in processes}

    while remaining or ready_queue:
        # Add arriving processes to ready queue
        arrived = [p for p in remaining if p[1] <= current_time]
        for p in arrived:
            ready_queue.append(p)
            remaining.remove(p)

        if not ready_queue:
            if remaining:
                current_time = remaining[0][1]
                continue
            else:
                break
        
        process = ready_queue.pop(0)
        pid, arrival, burst, priority = process
        
        start_time = current_time
        exec_time = min(burst_remaining[pid], quantum)
        current_time += exec_time
        burst_remaining[pid] -= exec_time
        
        segments.append((pid, start_time, current_time))
        
        # Check for arrivals during execution
        arrived = [p for p in remaining if p[1] <= current_time]
        for p in arrived:
            ready_queue.append(p)
            remaining.remove(p)
            
        if burst_remaining[pid] > 0:
            ready_queue.append(process)
        else:
            # Process finished
            completion_time = current_time
            tat = completion_time - arrival_times[pid]
            wt = tat - original_bursts[pid]
            pid_metrics[pid] = {'wt': wt, 'tat': tat}

    return segments, pid_metrics

def priority_scheduling(processes):
    # Non-preemptive priority scheduling
    processes.sort(key=lambda x: (x[1], x[3])) # Priority (lower value = higher priority usually)
    remaining = processes.copy()
    current_time = 0
    segments = []
    pid_metrics = {}

    while remaining:
        available = [p for p in remaining if p[1] <= current_time]
        if not available:
            current_time = min(p[1] for p in remaining)
            continue
        
        # Pick process with highest priority
        available.sort(key=lambda x: x[3])
        process = available[0]
        remaining.remove(process)
        
        pid, arrival, burst, priority = process
        start_time = current_time
        completion_time = start_time + burst
        waiting_time = start_time - arrival
        turnaround_time = completion_time - arrival
        
        segments.append((pid, start_time, completion_time))
        pid_metrics[pid] = {'wt': waiting_time, 'tat': turnaround_time}
        current_time = completion_time

    return segments, pid_metrics

File: web/src/utils/algorithms.js
export const calculateScheduling = (algorithm, processes, timeQuantum = 2) => {
    const procs = processes.map(p => ({
        pid: p.pid,
        arrival: parseInt(p.arrival),
        burst: parseInt(p.burst),
        priority: parseInt(p.priority) || 0
    }));

    switch (algorithm) {
        case "FCFS": return fcfs(procs);
        case "SJF": return sjf(procs);
        case "Round Robin": return roundRobin(procs, parseInt(timeQuantum));
        case "Priority Scheduling": return priorityScheduling(procs);
        default: return { segments: [], summary: {} };
    }
};

const fcfs = (processes) => {
    processes.sort((a, b) => a.arrival - b.arrival);
    let currentTime = 0;
    const segments = [];
    const metrics = {};

    processes.forEach(({ pid, arrival, burst }) => {
        if (currentTime < arrival) currentTime = arrival;
        const startTime = currentTime;
        const completionTime = startTime + burst;
        segments.push({ pid, start: startTime, end: completionTime });
        metrics[pid] = { wt: startTime - arrival, tat: completionTime - arrival };
        currentTime = completionTime;
    });

    return { segments, metrics };
};

const sjf = (processes) => {
    processes.sort((a, b) => a.arrival - b.arrival || a.burst - b.burst);
    const remaining = [...processes];
    let currentTime = 0;
    const segments = [];
    const metrics = {};

    while (remaining.length > 0) {
        const available = remaining.filter(p => p.arrival <= currentTime);
        if (available.length === 0) {
            currentTime = Math.min(...remaining.map(p => p.arrival));
            continue;
        }
        available.sort((a, b) => a.burst - b.burst);
        const process = available[0];
        const index = remaining.indexOf(process);
        remaining.splice(index, 1);

        const startTime = currentTime;
        const completionTime = startTime + process.burst;
        segments.push({ pid: process.pid, start: startTime, end: completionTime });
        metrics[process.pid] = { wt: startTime - process.arrival, tat: completionTime - process.arrival };
        currentTime = completionTime;
    }

    return { segments, metrics };
};

const roundRobin = (processes, quantum) => {
    processes.sort((a, b) => a.arrival - b.arrival);
    const remainingProcs = [...processes];
    const readyQueue = [];
    let currentTime = 0;
    const segments = [];
    const metrics = {};

    const burstRemaining = Object.fromEntries(processes.map(p => [p.pid, p.burst]));
    const originalBursts = Object.fromEntries(processes.map(p => [p.pid, p.burst]));
    const arrivalTimes = Object.fromEntries(processes.map(p => [p.pid, p.arrival]));

    while (remainingProcs.length > 0 || readyQueue.length > 0) {
        const arrived = remainingProcs.filter(p => p.arrival <= currentTime);
        arrived.forEach(p => {
            readyQueue.push(p);
            remainingProcs.splice(remainingProcs.indexOf(p), 1);
        });

        if (readyQueue.length === 0) {
            if (remainingProcs.length > 0) {
                currentTime = remainingProcs[0].arrival;
                continue;
            } else break;
        }

        const process = readyQueue.shift();
        const pid = process.pid;
        const startTime = currentTime;
        const execTime = Math.min(burstRemaining[pid], quantum);
        currentTime += execTime;
        burstRemaining[pid] -= execTime;

        segments.push({ pid, start: startTime, end: currentTime });

        const newlyArrived = remainingProcs.filter(p => p.arrival <= currentTime);
        newlyArrived.forEach(p => {
            readyQueue.push(p);
            remainingProcs.splice(remainingProcs.indexOf(p), 1);
        });

        if (burstRemaining[pid] > 0) {
            readyQueue.push(process);
        } else {
            const tat = currentTime - arrivalTimes[pid];
            metrics[pid] = { wt: tat - originalBursts[pid], tat };
        }
    }

    return { segments, metrics };
};

const priorityScheduling = (processes) => {
    processes.sort((a, b) => a.arrival - b.arrival || a.priority - b.priority);
    const remaining = [...processes];
    let currentTime = 0;
    const segments = [];
    const metrics = {};

    while (remaining.length > 0) {
        const available = remaining.filter(p => p.arrival <= currentTime);
        if (available.length === 0) {
            currentTime = Math.min(...remaining.map(p => p.arrival));
            continue;
        }
        available.sort((a, b) => a.priority - b.priority);
        const process = available[0];
        remaining.splice(remaining.indexOf(process), 1);

        const startTime = currentTime;
        const completionTime = startTime + process.burst;
        segments.push({ pid: process.pid, start: startTime, end: completionTime });
        metrics[process.pid] = { wt: startTime - process.arrival, tat: completionTime - process.arrival };
        currentTime = completionTime;
    }

    return { segments, metrics };
};
